Google account dialog no longer visible on send screen & account removal crash resolved

jd-alexander opened the pull request:
Closes #2479

#### What has been done to verify that this works as intended?

- Now when the account was set and then removed and you go to send a form it shows you the account selection dialog instead of crashing. 

- I turned the Wifi on and off so that a network event could be triggered that could cause the auto-send functionality to start and when it began and found there was no longer an account associated with the username that was stored it just stopped. 

- Now when an account is removed and you go to the Server settings instead of seeing the account listed the field is now blank since once the account has been removed and Collect is ran, it erases the value and returns an empty username. 

- Once you set an active account the functionality returns to normal and submissions can be made.

#### Why is this the best possible solution? Were any other approaches considered?
This is the best solution since it targets the main method that returning the user account and ensures the account exists before passing it off to the rest of the app.

#### Are there any risks to merging this code? If so, what are they?
No. We just have to verify that the functionality works as intended and there's no regression.

#### Do we need any specific form for testing your changes? If so, please attach one.
No. 

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lintDebug spotbugsDebug` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

jd-alexander committed:
Added a condition in getSelectedAccount that checks to see if account was deleted.

Now when code paths call to getSelectedAccount a check is done to see if the account is deleted. An account can be deleted for a variety of reasons. Sometimes a user might be signing into an org temporarily or a domain expires so the account is removed. So putting this in place ensures that no exceptions are thrown if this occurs which would cause an unwanted crash.

jd-alexander committed:
Checks to see if the google username exists before performing auto-send

Previously the google username was just being pulled from SharedPreferences. That wasn't reliable since as we have seen, the user can delete that account outside of Collect so utilize the GoogleAccountsManager to check for this ensures that the account is actually present.

jd-alexander committed:
Fixed a crash being caused by isEmpty being called on a null string

Also replaced the checks for null and isEmpty with the method from TextUtils since it does both.

lognaturel commented:
Looks good to me! You do have an unused import, though (see PMD report). Do make sure to either run pmd locally and see results or wait to see if the CircleCI build passes before checking the box in the PR template!

jd-alexander commented:
Yes, will do. I always run it locally. Gonna double check from now on!

jd-alexander committed:
fixed unused import issue.

jd-alexander committed:
Show a dialog to user that they need to use Settings to configure their Google account.

jd-alexander commented:
I'm gonna be refactoring some tests that were broken due to these changes and then I make a push for a review!

jd-alexander committed:
Added a null check for activity

jd-alexander committed:
Modified test structure to fit with the new approach being taken with getSelectedAccount

jd-alexander committed:
Reverted TextUtils.isEmpty() since it was calling JUnit tests to fail.

jd-alexander committed:
Refactored GoogleAccountsManagerTest

- Each time an account name is persisted to SharedPreferences an account with that name is added to the mocked Credentials to power the new getSelectedAccount function.

- Since there's now a check within chooseAccount to determine if a account chooser or a settings info should be displayed, the arguments for the condition were mocked in favour of showing the account chooser dialog

- Used lambda expressions to optimize some of the mocks.

jd-alexander commented:
@lognaturel quick question! I did a rebase to amend a commit message and the history got duplicated from the point of modification. Is that okay? I see a lot of discussions about preserving commit history but in this case since the PR hasn't been merged would it be good for me to find a way to get in without the duplicates?

jd-alexander committed:
Test to verify getSelectedAccount behaviour when Google account is deleted.

This test verifies the behaviour of the GoogleAccountsManager when an account is deleted from the device. This verifies that when the account is deleted that calling getSelectedAccount will cause the manager to return null which gets passed to other code paths that notify the user of the occurrence so that they can select another account.

lognaturel commented:
> history got duplicated from the point of modification

I'm sorry, I'm not sure what that means! I don't see any duplicated commits. Could you link to a duplicated pair? What I do see are merge conflicts that will need to be resolve.

What I'd probably do is squash commits 6a59172 to 4ae7984 since the compound change is easy to understand. I'd call that something like `Fix crash when Google account is deleted` (notice it's short so the summary which truncates at 70 chars will be completely shown on one line) and with the explanation you have at 6a59172. I would keep d030263 and ac8db95 but shorten the message for ac8db95 to something like `Test getSelectedAccount when Google account deleted` to avoid the wrapping.

> I see a lot of discussions about preserving commit history but in this case since the PR hasn't been merged would it be good for me to find a way to get in without the duplicates?

That's really important for `master` and for branches that are being collaborated on. This is up to some debate but I personally think it's totally ok to rebase and force push in an in-progress branch like this. Otherwise, once there are too many noisy commits like 37f4f84 or 4ae7984 then I have a tendency to squash the whole PR which is too bad.

shobhitagarwal1612 commented on collect_app/src/main/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManager.java:
```diff
@@ -170,7 +178,42 @@ public boolean checkAccountPermission() {
+        Account[] googleAccounts = credential.getAllAccounts();
+        String account = (String) preferences.get(PreferenceKeys.KEY_SELECTED_GOOGLE_ACCOUNT);
+
+        if (account == null) {
```
This condition is always `false`. The default value for this key is "" (blank) and not null
Line 106 : `PreferencesKeys.java`

shobhitagarwal1612 commented on collect_app/src/main/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManager.java:
```diff
@@ -170,7 +178,42 @@ public boolean checkAccountPermission() {
+                }
+            }
+
+            preferences.save(PreferenceKeys.KEY_SELECTED_GOOGLE_ACCOUNT, null);
```
Instead of saving `null`, we can reset this preference using the `reset` method in `GeneralSharedPreferences`. That way it will never be equal to null

jd-alexander replied:
Oh awesome! Never knew about that method.

shobhitagarwal1612 commented on collect_app/src/main/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManager.java:
```diff
@@ -170,7 +178,42 @@ public boolean checkAccountPermission() {
+
+            preferences.save(PreferenceKeys.KEY_SELECTED_GOOGLE_ACCOUNT, null);
+
+            return null;
```
redundant return statement

shobhitagarwal1612 commented on collect_app/src/test/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManagerTest.java:
```diff
@@ -66,24 +69,34 @@ private void stubAccountPermission(boolean permission) {
+    }
+
+    private void removeAccounts() {
+        doReturn(new Account[]{}).when(mockedCredential).getAllAccounts();
```
Method comment of `getAllAccounts()` says that 
```
 Returns all Google accounts or {@code null} for none. 
```

shobhitagarwal1612 replied:
So, returning null would be better in that case

jd-alexander replied:
Yeah! Best to match the signature for sure.

shobhitagarwal1612 commented on collect_app/src/main/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManager.java:
```diff
@@ -170,7 +178,42 @@ public boolean checkAccountPermission() {
+            return null;
+        }
+
+        return null;
```
We should return `""` here

shobhitagarwal1612 commented on collect_app/src/test/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManagerTest.java:
```diff
@@ -58,6 +60,7 @@
 
     private void stubSavedAccount(String accountName) {
         when(mockPreferences.get(PreferenceKeys.KEY_SELECTED_GOOGLE_ACCOUNT)).thenReturn(accountName);
+        stubAccount(accountName);
```
This stub can be removed as it is redundant and causes `credentials` to return blank account names which will never happen

jd-alexander replied:
Okay, I get your point. The reason I had to add this here is that the tests were failing when the `credentials` didn't contain the account being checked for within the `getSelectedAccount` method fix I implemented.

shobhitagarwal1612 replied:
Which test case is failing?

jd-alexander replied:
These tests were failing. 

![2018-08-18 23 03 19](https://user-images.githubusercontent.com/1509205/44305369-1e352f00-a33b-11e8-8acd-dc8ed0cc677f.png)

Yes, so mocking the `getAllAccounts` call allowed me to verify the new functionality adding in this PR and also verified that all the test cases relying on `getSelectedAccount` passed.

jd-alexander replied:
Also when I merged in the new permissions code #2474 it broke most of the tests so I had to figure out a way to mock the way how the action on the listener was called when the permission is granted.

shobhitagarwal1612 replied:
The new changes look good. But to ensure that credentials never return a blank account we should remove all `stubSavedAccount("");` from the test cases

lognaturel replied:
@jd-alexander there are still several `stubSavedAccount("");` in the tests. It would probably be best to change those to use a dummy account name, right?

lognaturel replied:
Actually, now I'm not so sure. It seems all those cases have to do with actually setting a blank account so it seems ok to me.

jd-alexander replied:
I just did a check and I realized that it's not necessary to set the blank account since `getSelectedAccount` is returning an empty string so the tests are able to verify functionality without utilizing a blank account which would be the opposite of how the underlying `GooogleAccountCredential' API operates so I am removing them and pushing as we speak.

lognaturel replied:
Ahhh, thank you, that makes sense. 😊

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into google_selected_account_issue

# Resolved conflicts:
#	collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java
#	collect_app/src/main/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManager.java
#	collect_app/src/main/res/values/strings.xml
#	collect_app/src/test/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManagerTest.java
- Performed mocking of PermissionsUtil
- Corrected the mock for GoogleAccountCredential's getAllAccounts
-  Performed other cleans relating to how the tests were interacting with mocked SharedPreferences

shobhitagarwal1612 commented on collect_app/src/main/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManager.java:
```diff
@@ -132,22 +141,53 @@ public void granted() {
         }
     }
 
     public String getSelectedAccount() {
```
This function can be annotated with NonNull

jd-alexander commented on collect_app/src/test/java/org/odk/collect/android/utilities/gdrive/GoogleAccountsManagerTest.java:
```diff
@@ -56,38 +66,51 @@
+
+    private void mockPermissionUtils() {
+        mockStatic(PermissionUtils.class, invocation -> {
+            Whitebox.invokeMethod(googleAccountsManager, "chooseAccount");
```
This was done because some of the approaches I was trying to utilize to invoke the direct `requestGetAccountsPermission` weren't working and I just decided that within this context the only function of the `PermissionsUtil` that we care about is this one so it's okay if an `Answer` just invokes the method that follows the permission being granted.

shobhitagarwal1612 replied:
Directly invoking the intended method smells like coupled code to me. What if `GoogleAccountManager` implements `PermissionListener` and then instead of invoking the method using reflection, we do,
```
PermissionListener permissionListener = googleAccountsManager();
permissionListener.onGranted()
```

jd-alexander replied:
Thats true! The reason why it had to be done like this is `PermissionUtils` contain static methods that encapsulate the permissions management provided by the `Dexter` library. The listeners aren't directly implemented on calling component so I am not able to get a referene to it as you described so that's the reason this approach was taken.

jd-alexander replied:
The design decisions around the permissions code made it easy to replicate the changes across the codebase in a simpler manner without having the calling being tangled to the permissions logic. That was the idea behind it.

jd-alexander committed:
Removed redundant code from GoogleAccountsManagerTest

jd-alexander commented:
@shobhitagarwal1612 ready for another review when you are! So we can get this merged in!

shobhitagarwal1612 commented on collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java:
```diff
@@ -134,14 +134,16 @@ private void uploadForms(Context context, boolean isFormAutoSendOptionEnabled) {
+                    GoogleAccountsManager accountsManager = new GoogleAccountsManager(Collect.getInstance());
+
+                    String googleUsername = accountsManager.getSelectedAccount();
                     if (googleUsername == null || googleUsername.isEmpty()) {
```
Null check can be eliminated

lognaturel replied:
@jd-alexander can you please remove this `null` check if it really isn't needed and then we should be ok to go to testing?

Agreed with @shobhitagarwal1612 we may want to rethink some of the way permissions are handled but I think we should do it elsewhere.

jd-alexander replied:
Yes, I thought I did this. Removing now. 

> Agreed with @shobhitagarwal1612 we may want to rethink some of the way permissions are handled but I think we should do it elsewhere.

Sure.

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into google_selected_account_issue

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into google_selected_account_issue

jd-alexander committed:
Removed null check on google username since it's never null.

lognaturel commented on collect_app/src/main/res/values/strings.xml:
```diff
@@ -653,6 +651,8 @@
     <string name="image_not_saved">The image can not be saved because of a problem with the corresponding instance file.</string>
     <string name="loading_form_failed">An error occurred while loading the form. Please try again.</string>
+    <string name="missing_google_account_dialog_title">Google account not set</string>
+    <string name="missing_google_account_dialog_desc">To submit forms via Google Sheets you have to setup your Google account in the Server settings.</string>
```
"setup" is a noun. In this case, we'd like the verb form which is "set up."

But it would be even better to reuse `google_invalid_account_title` and `google_invalid_account_description`. In general, we should be really careful about only adding new strings when absolutely necessary because they add a reading burden for enumerators and a translation burden for translators.

jd-alexander replied:
> because they add a reading burden for enumerators and a translation burden for translators.

That's a really good point. Will definitely keep that in mind when creating strings.

jd-alexander committed:
Removed duplicate google account strings

jd-alexander committed:
Stopped setting blank stub accounts since by design that's not how the underlying functionality works.

mmarciniak90 commented:
Hi @jd-alexander  
I was able to reproduce the crash on all devices.
The main case works as expected. The crash does not occur and the dialog appears.
Tested on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

But I noticed one small case which could be improved:
1. User sees dialog 'Google account not set'
2. User clicks on device back button
3. The dialog is closed but the user is not moved to the previous screen but sees an empty view. The user needs to clicks on device back button once more to backs to ODK. I feel that clicking on OK and device back button should cause the same effect in this case.

jd-alexander commented:
@mmarciniak90 thanks for spotting this! Gonna take a look!

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into google_selected_account_issue

jd-alexander committed:
Added an action to the onCancel listener that finishes the activity onBackPressed

jd-alexander commented:
@mmarciniak90  I resolved the issue and ran several tests myself and it's working great now. You can do a check and let me know if all is well.

mmarciniak90 commented:
@jd-alexander thanks for the quick fix! 

**Tested with success!**

Verified on Android 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1
* ODK Collect does not crash when Google Account is not longer available
* The dialog appears:
<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/45345011-67c51280-b5a5-11e8-87f8-7f1042d55632.png">

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

jd-alexander commented:
Awesome!
